src/ssh/cockpitsshrelay.c

```diff
@@ -1320,61 +1320,67 @@ send_auth_reply (CockpitSshData *data,
   return ret;
 }
 
-static void
+static gboolean
 parse_host (const gchar *host,
             gchar **hostname,
             gchar **username,
             guint *port)
 {
-  gchar *user_arg = NULL;
-  gchar *host_arg = NULL;
-  gchar *tmp = NULL;
-  gchar *end = NULL;
-
-  guint64 tmp_num;
-
-  gsize host_offset = 0;
-  gsize host_length = strlen (host);
-
-  tmp = strrchr (host, '@');
-  if (tmp)
-    {
-      if (tmp[0] != host[0])
+  GError *error = NULL;
+  g_autoptr (GRegex) regex = g_regex_new ("^"
+         "(?:(.+)@)?"         /* optional username */
+         "(?|"                /* one of... */
+           "\\[([^]@]+)\\]"     /* hostname in square brackets, no @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@:]+)"           /* hostname with no : or @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@]+)"            /* hostname with no @ but : (IPv6 address), and no port */
+         ")"                  /* . */
+       "$",
+       0, 0, &error);
+  g_assert_no_error (error);
+
+  g_autoptr(GMatchInfo) info = NULL;
+
+  if (g_regex_match (regex, host, 0, &info))
+    {
+      g_autofree gchar *port_str = g_match_info_fetch (info, 3);
+      gchar *endptr;
+      gint64 value;
+
+      if (port_str)
         {
-          user_arg = g_strndup (host, tmp - host);
-          host_offset = strlen (user_arg) + 1;
-          host_length = host_length - host_offset;
+          value = g_ascii_strtoull (port_str, &endptr, 10);
+          if (port_str[0] == '\0' || *endptr != '\0' || value > G_MAXUINT16)
```

Comment:
...and remove the `port_str[0] == '\0'` check here plz.  fwiw, it's not possible for the port to be empty in the case you're probably thinking about here, because we match on `[0-9]+`.

In general, we probably need to do a little bit less error checking here, since we know that the string in question is definitely made of all digits, with no whitespace, and not negative.  The only real possible error is that it's too large...  You could almost just do `atoi()` and check the range.  On the other hand, the extra checking doesn't really hurt, either.

Reply:
Good point -- I turned the endptr check into an assertion.